### PR TITLE
Bug 1830348 - Use do not intercept cache in app links interceptor

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1830348

### CODE REVIEW

File: android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt [hunk 1]
```diff
@@ -295,11 +295,46 @@ class AppLinksUseCases(
...
+            val currentTimeStamp = SystemClock.elapsedRealtime()
+
+            if (cacheTimeStamp != null &&
+                currentTimeStamp <= cacheTimeStamp + APP_LINKS_DO_NOT_OPEN_CACHE_INTERVAL
+            ) {
+                return true
+            }
+
+            return false
+        }
```
Comment (t-p-white) [inline android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:329]: drive-by - could this be something like:
```suggestion
return cacheTimeStamp != null && currentTimeStamp <= (cacheTimeStamp + APP_LINKS_DO_NOT_OPEN_CACHE_INTERVAL)
```

Comment (rocketsroger) [reply android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt]: Good catch.  Updated, thanks.

File: android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt [hunk 2]
```diff
@@ -295,11 +295,41 @@ class AppLinksUseCases(
...
+            return cacheTimeStamp != null &&
+                currentTimeStamp <= (cacheTimeStamp + APP_LINKS_DO_NOT_OPEN_CACHE_INTERVAL)
+        }
+
+        internal fun addUserDoNotIntercept(url: String, appIntent: Intent?) {
+            val cacheKey = getCacheKey(url, appIntent)
+            cacheKey?.let {
+                userDoNotInterceptCache[it] = SystemClock.elapsedRealtime()
+            }
+        }
```
Comment (titooan) [inline android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:331]: I think those functions can live directly into AppLinksUseCases instead of the Companion object, and be called from `AppLinksFeatures` and `AppLinksInterceptor` using the instance of `appLinksUseCases` they already have. What do you think?

Comment (rocketsroger) [reply android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:331]: Unfortunately `userDoNotInterceptCache` needs to be in companion.  Since `AppLinksFeature` and `AppLinksInterceptor` doesn't use the same instance of `AppLinksUseCases`.  So moving these functions outside of the companion might be confusing, since we don't need an instance of `AppLinksUseCases` to access the cache.  We should allow the user to access it without creating an instance.

**Review by titooan — APPROVED** [review]
LGTM!

**Review by titooan — APPROVED** [review]

File: android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt (hunk 2 above)
Comment (rocketsroger) [reply android-components/components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt]: As discussed, I've moved the cache into `AppLinksInterceptor` . Thanks giving me this idea!
